include/envoy/network/connection.h

```diff
@@ -326,6 +326,8 @@ class Connection : public Event::DeferredDeletable,
    *  returned.
    */
   virtual absl::optional<std::chrono::milliseconds> lastRoundTripTime() const PURE;
+
+  virtual OptRef<const Network::Socket> socket() const PURE;
```

Comment:
I think if we're going to do this we'd probably want the socket to be const, WDYT?

Reply:
It is, do you mean the OptRef to be const too?

Reply:
I did, but as you say it's possibly overkill.
fine as-is, but you need to check format ci?

/wait

Reply:
right, the format is done

Reply:
A drive by comment, what is rational behind using  `OptRef` and not a shared ptr to a socket or a refence to a shared ptr (we do the latter for `socketOptions()`)?

Reply:
It could be a shared ptr but not a reference, because It can be null in some implementations, it depends of style and I was trying to avoid possible nullptr exceptions, do you suggest use a shared ptr instead?

Reply:
Not a suggestion, I just trying to understand what the correct pattern is myself. This is why I commented.

I think that returning an optional reference is used way more widely.

I see two different patterns used:
1. `OptRef<T>`
2. `absl::optional<std::reference_wrapper<T>>;`, we also have an existing using `using SocketOptRef = absl::optional<std::reference_wrapper<Socket>>;`

I am not sure which one is more preferable? Should the second one be refactored to use `OptRef<>`.